iib/workers/tasks/build.py

```diff
@@ -337,6 +340,151 @@ def _get_resolved_image(pull_spec):
+
+    try:
+        data = json.loads(
+            run_cmd(
+                ['podman', 'image', 'inspect', from_index], exc_msg='Failed to fetch image metadata'
+            )
+        )
+        db_path = data[0]['Config']['Labels']['operators.operatorframework.io.index.database.v1']
+        run_cmd(
+            ['podman', 'cp', f'{cid}:{db_path}', base_dir],
+            exc_msg='Failed to copy DB file from the container',
+        )
+    finally:
+        run_cmd(['podman', 'rm', cid], exc_msg='Failed to delete the container')
+    local_path = os.path.join(base_dir, db_path.split('/')[-1])
+
+    return local_path
+
+
+def _serve_index_registry(db_path):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str db_path: path to index database containing the registry data
+    :return: tuple containing port number of the running service and the runnung Popen object
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason
+    """
+    conf = get_worker_config()
+    port = conf['iib_grpc_start_port']
+    for i in range(conf['iib_grpc_max_port_tries']):
+        cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+        rpc_proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+        )
+        time.sleep(conf['iib_grpc_init_wait_time'])
+        ret = rpc_proc.poll()
+        if not ret:
+            log.debug('Running the command "%s"', ' '.join(cmd))
+            break
+        else:
+            stderr = rpc_proc.stderr.read()
+            if 'address already in use' in stderr:
+                if i == conf['iib_grpc_max_port_tries'] - 1:
+                    err_msg = (
+                        'No free port has been found after %s attempts.'
+                        % conf['iib_grpc_max_port_tries']
+                    )
+                    log.error(err_msg)
+                    raise IIBError(err_msg)
+                log.info('Port %s is in use, trying another.', port)
+                port += 1
+            else:
+                raise IIBError('Command "%s" has failed with error "%s"', ' '.join(cmd), stderr)
+
+    return (port, rpc_proc)
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
+    """
+    Get a list of bundles already present in the index image.
+
+    :param str from_index: index image to inspect
+    :param str base_dir: base directory to create temporary files in
+    :return: list of present bundles as provided by the grpc query
+    :rtype: list
+    :raises IIBError: if any of the commands fail
+    """
+    db_path = _get_index_database(from_index, base_dir)
```

Comment:
The container itself serves the DB, so just running it would suffice. Also, I think that grpcurl isn't installed in the container, so it would have to be run locally. 

In any case, I believe Matt was against running a container in IIB to avoid arbitrary code execution on IIB worker. If nothing else, running the DB directly has probably less overhead than running a container. Also, if we ever decide to run IIB in containers (as we do with dev-env), running another container in it might be hard to achieve (I think).